Resolve jmethodID issues with class strong reference cache

jbachorik opened the pull request:
This is the last attempt to resolve the issues related to the specification and implementation of `jmethodID` type.

## Problem
In short, there is no 'safe' way to use the jmethodID values unless the caller is holding a strong reference to the the class containing the method that particular jmethodID is pointing to. This makes the delayed processing of stacktraces potentially dangerous as the classes referenced from a stacktrace might have been unloaded in the meantime.

There is quite a lot of specific code in JVM which is trying to make sure that once a class gets unloaded all jmethodID values corresponding to the methods of such class would become invalid (eg. by NULL-ing) and it should be possible to check for validity and recover gracefully if the method metadata is not available any more. However, in practice this is failing randomly due to very complex interplay of the GC, metaspace handling and JVM class retransformation.

Empirically we confirmed that all the observed crashes are related to stacktraces collected via JVMTI and the stacktraces collected via ASGCT are somehow resilient to this problem.

## Class-Reference Cache

After previous two attempts on preemptively caching all method metadata just in case some classes get unexpectedly unloaded, which lead to underperforming solution, this is the last attempt to workaround the jmethodID issue.
The gist of the workaround is to create strong global JNI references as soon as we obtain the stacktrace from JVMTI, thus preventing the classes being unloaded for as long as we hold the references. Since we need the classes to survive until the next recording flush only, we can safely release all references when we are finalizing that recording.

The concurrent nature of stack allocations requires a double-buffered cache - when we enter the finalization process of the recording we immediately switch the used cache instance such that all allocations happening during the finalization will go to that instance instead of the one being released. This is to avoid releasing references prematurely.

Another quirk we need to take care of is the live heap profiling samples - the related stacktraces can survive multiple finalization cycles and as such can not be immediately released. Therefore, we are keeping a set of references coming from the live heap samples and the finalization process will skip such references. This second reference set is updated each time the table of the live samples is updated and during the update all old references are released, unless they are in the new reference set.

## jmethodID Validity Check
Unfortunately, even with the class-reference cache it was not possible to avoid all crashes, only the most of them. Further investigation showed that the remaining crash is related to recording-on-shutdown only - apparently, JVM starts aggressively unloading classes, regardless of the JNI strong references. Good news is that this is not a concurrent class unloading process so once we check jmethodID for validity it will stay valid and we can safely use it in JVMTI calls.
The validity check is created using the 'vmStructs' to obtain offsets of the fields the JVM code is accessing when resolving a particular jmethodID to method metadata. As such it relates on the `vmStructs` to be available and up-to-date but it is the best we have.

---
### Note!
The PR is split into self-standing commits - I had to remove the old jmethodID cache implementation and class-unload hook but I didn't feel like it made sense to have them as self-standing PRs just to be followed by this one.
I think the commits are pretty concise but if there is a need I might split this PR into separate commits.

richardstartin commented on ddprof-lib/src/main/cpp/classRefCache.cpp:
```diff
@@ -0,0 +1,72 @@
+        epoch = _epoch;
+    } while (!__sync_bool_compare_and_swap(&_epoch, epoch, epoch == 0 ? 1 : 0));
+
+    return ClearTask(epoch, force);
```
A copy will be made here, so the destructor will be invoked when the stack pops, which will do a clean. I think it would be better to rework this so that the epoch is acquired by this method, and the caller stack-allocates the RAII `ClearTask` as and when they need to use it. More boilerplate but clearer ownership.

jbachorik replied:
Ok, I have modified the code.

richardstartin commented on ddprof-lib/src/main/cpp/flightRecorder.cpp:
```diff
@@ -367,6 +368,8 @@ off_t Recording::finishChunk(bool end_recording) {
 
 void Recording::switchChunk(int fd) {
+    // ClearTask is a RAAI that will switch the epoch immediately and release the cached references when going out of scope
+    ClearTask cc = ClassRefCache::instance()->clear();
```
See above, I think the `ClearTask` needs to be constructed from an acquired epoch here to achieve the desired semantics.

richardstartin reviewed (changes requested):
Mostly looks good, but I think the comments relating to `ClearTask` need to be addressed - should be a very easy change to accommodate this though.

szegedi commented on ddprof-lib/src/main/cpp/vmEntry.h:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright 2016 Andrei Pangin
+ * Copyright 2021, 2023 Datadog, Inc
```
If we aren't adding anything to the file, should we put a copyright notice on it?

jbachorik replied:
Yeah, this is catch-up .. it should have been updated with the previous changes. Now they are rolled back so this looks like a no-op.

jbachorik commented on ddprof-lib/src/main/cpp/flightRecorder.cpp:
```diff
@@ -110,27 +115,81 @@ void Lookup::cutArguments(char* func) {
-    mi->_type = cmi->_type;
-    mi->_is_entry = cmi->isEntry();
+        // Check if the frame is Thread.run or inherits from it
+        if (strncmp(method_name, "run", 3) == 0 && strncmp(method_sig, "()V", 3) == 0) {
```
This is an inherited code from async-profiler, but your suggestion makes sense.
Thanks!

jbachorik commented on ddprof-lib/src/main/cpp/arch.h:
```diff
@@ -37,6 +37,10 @@ static inline int atomicInc(volatile int& var, int increment = 1) {
+static inline int loadAcquire(volatile int& var) {
+    return __atomic_load_n(&var, __ATOMIC_ACQUIRE);
+}
+
```
Thanks, will reorder.
I will leave the 'templatization' for some time later, I don't want to pollute the PR.

richardstartin commented on ddprof-lib/src/main/cpp/flightRecorder.cpp:
```diff
@@ -143,13 +143,8 @@ void Lookup::fillJavaMethodInfo(MethodInfo* mi, jmethodID method, bool first_tim
-            if (class_modifiers & 0x0040) {
-                modifiers |= 0x0040;
-            }
+            // in order to be able to identify 'hidden' frames the "SYNTHETIC" (0x1000) and "BRIDGE" (0x0040) modifiers will be propagated to methods
+            modifiers |= class_modifiers & (0x1000 | 0x0040);
```
(bikeshed) maybe pull these masks into constants and replace with `modifiers |= class_modifiers & (SYNTHETIC | BRIDGE);` and remove the comment

danielsn commented on ddprof-lib/src/main/cpp/vmStructs.cpp:
```diff
@@ -516,6 +517,44 @@ jmethodID VMMethod::id() {
+ * This check is attempting to validate that all data structures required to reconstruct the method metadata
+ * from a jmethodID are still readable at this point.
+ */
+bool VMMethod::check_jmethodID(jmethodID id) {
```
This check appears to be a best-effort. Maybe note that in the name?

jbachorik commented on ddprof-lib/src/main/cpp/classRefCache.cpp:
```diff
@@ -0,0 +1,87 @@
+    for (auto const& clz : classes) {
+        refs.insert(jni->NewGlobalRef(clz));
+    }
+    int epoch = loadAcquire(_epoch);
+    _cache_lock[epoch].lock();
```
The `_epoch` is changing only when recording is finalized, eg. by default each 60s.
The implementation trades simplicity for correctness only within the limited boundaries. Eg. we assume that there will not be several 10s of seconds between acquiring the epoch and locking the corresponding lock.

jbachorik commented on ddprof-lib/src/main/cpp/flightRecorder.cpp:
```diff
@@ -110,27 +115,81 @@ void Lookup::cutArguments(char* func) {
+        }
+        deallocateStrings = true;
+    } else {
+        class_name = (char*)"L;";
+        method_name = (char*)"jvmtiError";
+        method_sig = (char*)"()L;";
```
AFAIK, this is safe. At least, the same code has been there since beginning.

r1viollet replied:
I would say this is a style discussion.
The type is here to help. You would usually avoid c-style casts in favour of the appropriate c++ cast. And you usually would avoid using a const_cast. I think you introduced this to unify the code paths. I probably would have added an extra variable to avoid this.

jbachorik commented on ddprof-lib/src/main/cpp/classRefCache.cpp:
```diff
@@ -0,0 +1,87 @@
+    do {
+        epoch = _epoch;
+    } while (!__sync_bool_compare_and_swap(&_epoch, epoch, epoch == 0 ? 1 : 0));
+
```
What exactly do you mean by that?
This will atomically switch 1->0 and 0->1 - what would be the exact flow that would cause problems with `_epoch` changing?

danielsn replied:
Same question as before: if another thread tried to do this too, we would have a conflict.  Maybe the answer is that this happens only if it takes >60 seconds to do the swap, so its sufficiently improbably as to not be a concern.

jbachorik replied:
I see. Yes, this is all based on the assumption that the recording 'flush' is not happening from more than 2 threads concurrently.
Perhaps, the CAS loop is not even necessary here because of that assumption.

jbachorik commented on ddprof-lib/src/main/cpp/vmEntry.h:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright 2016 Andrei Pangin
+ * Copyright 2021, 2023 Datadog, Inc
```
I explained that - the copyright should have been updated in the previous PR, the changes were rolled back so this seems like a no-op now.

jbachorik commented on ddprof-lib/src/main/cpp/flightRecorder.cpp:
```diff
@@ -110,27 +115,81 @@ void Lookup::cutArguments(char* func) {
+        method_sig = (char*)"()L;";
+    }
+
+    mi->_class = _classes->lookup(class_name + 1, strlen(class_name) - 2);
```
This is a legacy code, just moving around.
But in short, it is based on the class descriptor format in Java - eg. `Ljava/lang/String;` where the non-primitive type is introduced by `L` and is ended by `;`

jbachorik commented on ddprof-lib/src/main/cpp/flightRecorder.cpp:
```diff
@@ -110,27 +115,81 @@ void Lookup::cutArguments(char* func) {
+        if (jvmti->GetClassModifiers(method_class, &class_modifiers) == 0 && jvmti->GetMethodModifiers(method, &modifiers) == 0) {
+            // class constants are written without the modifiers info
+            // in order to be able to identify 'hidden' frames the "SYNTHETIC" and "BRIDGE" modifiers will be propagated to methods
+            if (class_modifiers & 0x1000) {
```
Not any more ...

jbachorik commented:
Thanks everyone for the reviews.
Unless someone feels the unaddressed comments are blocking, let me know. Otherwise I am going to queue this PR for merge in ~1hr.

jbachorik commented:
/merge

jbachorik commented on ddprof-lib/src/main/cpp/livenessTracker.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright 2021, 2022 Datadog, Inc
+* Copyright 2021, 2023 Datadog, Inc
```
I think so, but it's not like I saw any guidelines. Just applying what I was doing at Oracle and they can be pretty sensitive about copyright :)
